contracts/zero-ex/contracts/src/features/MetaTransactionsFeatureV2.sol

```diff
@@ -0,0 +1,620 @@
+                .rrevert();
+        }
+        returnResults = new bytes[](mtxs.length);
+        for (uint256 i = 0; i < mtxs.length; ++i) {
+            ExecuteState memory state;
+            state.sender = msg.sender;
+            state.mtx = mtxs[i];
+            state.hash = getMetaTransactionV2Hash(mtxs[i]);
+            state.signature = signatures[i];
+
+            returnResults[i] = _executeMetaTransactionPrivate(state);
+        }
+    }
+
+    /// @dev Get the block at which a meta-transaction has been executed.
+    /// @param mtx The meta-transaction.
+    /// @return blockNumber The block height when the meta-transactioin was executed.
+    function getMetaTransactionV2ExecutedBlock(
+        MetaTransactionDataV2 memory mtx
+    ) public view override returns (uint256 blockNumber) {
+        return getMetaTransactionV2HashExecutedBlock(getMetaTransactionV2Hash(mtx));
+    }
+
+    /// @dev Get the block at which a meta-transaction hash has been executed.
+    /// @param mtxHash The meta-transaction hash.
+    /// @return blockNumber The block height when the meta-transactioin was executed.
+    function getMetaTransactionV2HashExecutedBlock(bytes32 mtxHash) public view override returns (uint256 blockNumber) {
+        return LibMetaTransactionsStorage.getStorage().mtxHashToExecutedBlockNumber[mtxHash];
+    }
+
+    /// @dev Get the EIP712 hash of a meta-transaction.
+    /// @param mtx The meta-transaction.
+    /// @return mtxHash The EIP712 hash of `mtx`.
+    function getMetaTransactionV2Hash(MetaTransactionDataV2 memory mtx) public view override returns (bytes32 mtxHash) {
+        return
+            _getEIP712Hash(
+                keccak256(
+                    abi.encode(
+                        MTX_EIP712_TYPEHASH,
+                        mtx.signer,
+                        mtx.sender,
+                        mtx.expirationTimeSeconds,
+                        mtx.salt,
+                        keccak256(mtx.callData),
+                        mtx.feeToken,
+                        mtx.fees
+                    )
+                )
+            );
+    }
+
+    /// @dev Execute a meta-transaction by `sender`. Low-level, hidden variant.
+    /// @param state The `ExecuteState` for this metatransaction, with `sender`,
+    ///              `hash`, `mtx`, and `signature` fields filled.
+    /// @return returnResult The ABI-encoded result of the underlying call.
+    function _executeMetaTransactionPrivate(ExecuteState memory state) private returns (bytes memory returnResult) {
+        _validateMetaTransaction(state);
+
+        // Mark the transaction executed by storing the block at which it was executed.
+        // Currently the block number just indicates that the mtx was executed and
+        // serves no other purpose from within this contract.
+        LibMetaTransactionsStorage.getStorage().mtxHashToExecutedBlockNumber[state.hash] = block.number;

```

Comment:
If this isn't used elsewhere, we should get rid of it. Penalizing the relayer an extra 20k gas per transaction is hefty

Reply:
Other than storing information in a map, how else would we know that a metatransaction has been executed? It's important you can't execute the metatransaction again, since these are being submitted by a third party on the signer's behalf.